### ISSUE TITLE
frida-server crashes Anbox when rootfs-overlay is enabled

### ISSUE THREAD

**Issue by baptx** [frida/frida#1187]
You can reproduce the problem by installing the latest Anbox snap package with this command (tested with beta and edge):
`snap install --beta --devmode anbox`
Then enable rootfs-overlay in Anbox and restart it with these commands (this is necessary to add a certificate from a tool like mitmproxy in `/var/snap/anbox/common/rootfs-overlay/system/etc/security/cacerts/` if we want to intercept HTTPS requests, since `/system/` folder is read-only):
`snap set anbox rootfs-overlay.enable=true`
`snap restart anbox`
Start Anbox app manager with this command:
`anbox.appmgr`
Then add latest frida-server to Anbox and start it:
`adb push frida-server /data/local/tmp/`
`adb shell "chmod 755 /data/local/tmp/frida-server"`
`adb shell "su 0 /data/local/tmp/frida-server -l 0.0.0.0" &`

After frida-server is started and we try to start an app through the app manager GUI, for example the WebView Browser Tester, Anbox crashes. It crashes even if rootfs-overlay folder is empty. However it does not crash when rootfs-overlay is not enabled. The problem did not happen in the past.
I reinstalled Anbox today on Lubuntu 19.10 and with the command `snap info anbox`, I can see that the latest beta version is from 2020-01-02. I am also using the latest version of Frida on the host (12.8.11) and on the target (frida-server-12.8.11-android-x86_64).
The same version of frida-server works great on Android x86_64 7.1-r2 in VirtualBox.

Do you know where the problem comes from and how to fix it?

I created the same issue on Anbox project since I don't know exactly where the problem comes from: https://github.com/anbox/anbox/issues/1368

**Comment by liornm** [comment]
Hey!

After debugging this issue today I have discovered that frida-agent-32.so causes zygote to crash for some reason. After zygote crashes Anbox immediately crashes as well.

Because I am lazy, I have decided to look for a workaround instead of fixing this bug. After trial and error I have found that this bug is not present in the 64 bit version of the agent (frida-agent-64.so). Therefore, after disabling 32 bit agent Frida can still inject and execute every 64 bit or libhoudini (arm/arm64) app.

In order to prevent frida-server from injecting the agent to zygote (32 bit) I have simply modified two lines, see [the last commit](https://github.com/liornm/frida-core/commit/7a5d8ffcad7209f5e28265ae360504a6a09c96b2).

If you want to avoid the hassle of compiling frida-server yourself you can just use my [compiled version](https://github.com/liornm/frida-core/releases/tag/anbox-workaround).


Let me know if it works for you.

Cheers.

**Comment by baptx** [comment]
@liornm hi, thanks for your reply. Will your fix be added to the official Frida version soon? I will test next time when I need to use Frida if I have a moment.

**Comment by liornm** [comment]
It is not a fix, but a workaround, so it is not going to be merged to the mainline any time soon.
You can find the modified line the master branch here: [linux-host-session.vala:584](https://github.com/frida/frida-core/blob/55e7d13915e03fc4a9d376d601a17f9889da7737/src/linux/linux-host-session.vala#L584).

**Comment by baptx** [comment]
@liornm thanks, I guess I did not find the line yesterday because I searched `cmdline == "zygote" || ` instead of `name == "zygote" || ` in `linux-host-session.vala`.
Do you think the bug can be fixed from the Anbox side? Or it is just a Frida bug?

**Comment by liornm** [comment]
@baptx I am not so sure.
I believe the right way to go would be fixing Frida.

**Comment by baptx** [comment]
@liornm I took a moment to test your workaround on the latest Frida version 15.1.17 but it did not fix the issue (I tested on Ubuntu 20.04 and the beta channel of Anbox which is more stable since the edge channel did not work when I tried). On my computer, I installed the latest Frida version with the command `sudo pip install frida-tools`.

Here are the steps I followed in case someone wants to try to build frida-server:
`git clone --recurse-submodules https://github.com/frida/frida` (like mentioned on https://frida.re/docs/building/#gnulinux).
`cd frida`
`git checkout tags/15.1.17` (to match the latest stable version like mentioned on https://stackoverflow.com/questions/4330610/switch-to-another-git-tag/4330627#4330627)

Then I downloaded Android NDK r24 on https://developer.android.com/ndk/downloads which was required for compiling and used it with this command (replace the value of the environment variable with the appropriate path):
`export ANDROID_NDK_ROOT=/path/to/android-ndk-r24`

Then I edited the 2 files necessary  for the workaround and executed the following command:
`make core-android-x86_64`

If I understand correctly, even if this workaround would work, we would be limited to 64-bit apps.

@oleavr Would it be possible to have a proper / official fix?
